### PR TITLE
Add tests for interaction between import() and microtask queue

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
The HTML algorithms for dynamic import (starting from [HostImportModuleDynamically](https://html.spec.whatwg.org/#hostimportmoduledynamically(referencingscriptormodule,-modulerequest,-promisecapability))) only schedule a task a when actually fetching the corresponding files, and it never happens if the imported module (and all its dependencies) have already been loaded (when step 5 of [fetch a single module script](https://html.spec.whatwg.org/#fetch-a-single-module-script) doesn't happen and step 6 returns early).

I didn't know how to test "it doesn't schedule a task", so the tests are written using a loop that schedules many microtasks and checking if all those microtasks are run or not before that the `import()` promise is resolved.

---

Browsers implement this inconsistently. This PR adds three new tests, and some of them fail:
| *Test* | *Firefox* | *Chrome* | *Safari* |
|:-|:-:|:-:|:-:|
| import() should not drain the microtask queue if it fails during specifier resolution | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| import() should not drain the microtask queue if it fails while validating the 'type' assertion | :grey_question: | :heavy_check_mark: | :heavy_check_mark: [^1] |
| import() should not drain the microtask queue when loading an already loaded module | :x: | :x: | :heavy_check_mark: |
| import() should drain the microtask queue when fetching a new module | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |

[^1]: The fix has not been released yet, so it fails on CI.

It's possible that the Firefox/Chrome behavior is because the HTML spec uses a rare "asynchronous completions" wording which isn't defined anywhere. However, "asynchronous completions" are used for all the test cases, so I would expect the 1st to always behave like the 2nd (since they both only involve asynchronous completions and no fetching). I also opened https://github.com/whatwg/html/pull/8264 to move away from that wording.

### CODE REVIEW

**Comment by domenic** [comment]
So I think we have two consistent options:

1. Always schedule a full task (i.e., always drain the microtask queue) to resolve the import() promise
2. Always synchronously resolve the import() promise if possible (i.e. if no network is involved)

It sounds like Safari follows option (2), and no browser follows option (1). So a path of least resistance would be to spec (2) and update Chrome/Firefox.

@hiroshige-g, what do you think from Chrome's perspective?

**Comment by nicolo-ribaudo** [comment]
For context, I discovered this behavior when working on https://github.com/tc39/ecma262/pull/2905 / https://github.com/whatwg/html/pull/8253.
With that refactor, ECMA-262 already knows what the result of the second dynamic import call will be and could potentially avoid calling the host hook again (especially given that the host hook is required to "return" the same result). However, I had to keep that unnecessary HTML<->ECMA-262 call so that hosts can decide _when_ to "return" and potentially inject an event loop spin.

**Comment by hiroshige-g** [comment]
When we implemented module loading, I interpreted the "asynchronous completion" in the spec as something like posting a task to make it always async, while  https://github.com/whatwg/html/pull/8264 made it explicitly synchronous in some cases. What was the intention of the "asynchronous completion"?

So the Chromium implementation is roughly:

(3) Always schedule at least one full task (i.e., always drain the microtask queue) for the fetch * module graph algorithms

This is consistent with the test result. import() with invalid URL is rejected synchronously because module specifier resolution is outside the fetch * module graph algorithms.

**Comment by nicolo-ribaudo** [comment]
Note that URL resolution/validation happens inside _fetch an import() module script graph_, and before https://github.com/whatwg/html/pull/8264 HostImportModuleDynamically called that algorithm using the "asynchronous completions" pattern (so it should have been consistent with the other parts of _fetch ..._ that do not involve network requests).

**Comment by domenic** [comment]
> What was the intention of the "asynchronous completion"?

I can't really say :(. I was bad and didn't think properly about it.

> import() with invalid URL is rejected synchronously because module specifier resolution is outside the fetch * module graph algorithms.

As @nicolo-ribaudo, I don't think that's correct. HostImportModuleDynamically calls fetch an import() module script graph, which if resolution fails, used the "asynchronously complete this algorithm with null" wording.

After https://github.com/whatwg/html/pull/8264, it uses sync wording, but we're wondering what the best path is...

Re-reading the algorithms I noticed we have a number of early-failure modes, actually:

- use inside WorkletGlobalScope and ServiceWorkerGlobalScope
- module specifier resolution failure
- moduleType is not "css", "json", or "javascript"
- moduleType is "css" and we're inside a WorkerGlobalScope

We should probably be consistent with all of these? (And ideally even test all of them.)

So generalizing, I think the options on the table are:

1. Always schedule a full task, including for early-success and early-failure
2. Always perform early-success and early-failure synchronously
3. (Chrome's current behavior, probably) Perform early-success with a full task, and early-failure synchronously

Reiterating my earlier comment, Chrome's current behavior feels inconsistent to me so I would prefer (1) or (2). And Safari matches (2).

**Comment by nicolo-ribaudo** [comment]
I added a test for invalid `type` assertions, and noticed that Chrome handles that error differently from the invalid URL error.

**Comment by codehag** [comment]
Thanks for this @nicolo-ribaudo. Option 2 makes the most sense to me, I don't think there will be an issue updating our behavior to match. I think our interpretation of the asynchronous completions was the similar to hiroshige's, but jonco can confirm if that is true. ccing @jonco3 and @smaug---- to get a couple more eyes from mozilla on this.

**Comment by nicolo-ribaudo** [comment]
Now that https://bugs.webkit.org/show_bug.cgi?id=246207 has been fixed, the Safari column should be all :heavy_check_mark:

File: html/semantics/scripting-1/the-script-element/module/dynamic-import/microtasks/basic.html [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+<script type="module">
+import { ticker } from "./ticker.js";
+
+promise_test(async t => {
+  const getCount = ticker(1000);
+
+  try {
+    await import("<invalid>");
+    assert_unreached('import() should reject');
+  } catch (_) {}
```
Comment (domenic) [inline html/semantics/scripting-1/the-script-element/module/dynamic-import/microtasks/basic.html:14]: Use https://web-platform-tests.org/writing-tests/testharness-api.html#promise-rejection for these.

**Review by domenic — COMMENTED** [review]
These look great modulo a minor fix, and they test the desired outcome.

It would be ideal if you also expanded to test:

- moduleType is "css" and we're inside a WorkerGlobalScope
- use inside WorkletGlobalScope
- use inside ServiceWorkerGlobalScope

but of course we're happy to accept just these too.

I believe the spec as written also aligns with these tests, correct? In which case we can merge this shortly.

To fix CI failures, often rebasing on master works.

**Comment by nicolo-ribaudo** [comment]
Is there a way to say "run this test in a worker/etc", or do I have to write a normal test that spins up a worker/etc?

**Comment by domenic** [comment]
The easiest way is to use [`.any.js` tests](https://web-platform-tests.org/writing-tests/testharness.html?highlight=worker#tests-for-other-or-multiple-globals-any-js) with a custom `global=` configuration.

Worklets are harder. I guess maybe modify the test I added in https://github.com/web-platform-tests/wpt/commit/43fb3b28428eaa6a5eb59fb4fda4b827bf1dced7 .

**Comment by nicolo-ribaudo** [comment]
I'm not sure about what to do with the CI failure, it looks like the basic tests sometimes timeout in Firefox.

**Review by domenic — APPROVED** [review]
LGTM. The CI is often flaky (see https://github.com/web-platform-tests/wpt/issues/21706) and so in these cases we summon one of the admins to help force-merge... I'll work on that.

File: html/semantics/scripting-1/the-script-element/module/dynamic-import/microtasks/worklet.https.html [hunk 2]
```diff
@@ -0,0 +1,43 @@
+<!DOCTYPE html>
+<html class="reftest-wait">
+<link rel="help" href="https://html.spec.whatwg.org/#hostimportmoduledynamically(referencingscriptormodule,-specifier,-promisecapability)">
+<link rel="match" href="worklet-ref.html">
```
Comment (nicolo-ribaudo) [inline html/semantics/scripting-1/the-script-element/module/dynamic-import/microtasks/worklet.https.html:4]:
```suggestion
<link rel="match" href="worklet-ref.https.html">
```

**Pull request merged**